### PR TITLE
HQO for scale/zero point

### PR DESCRIPTION
**Pull request by Giuseppe5**
Implement HQO optimization for scale and zero point.
Caveats:
- Scale does not seem to perform that well
- Zero point supports per_tensor/per_channel/per_group (asym only)
- Scale supports per_tensor/per_channel
- More numerical tests are needed to evaluate the impact of HQO compared to other techniques
- Scale implementation is INT-only
- Zero point implementation does not make sense for FP8 since we mostly do symmetric quantization

Maybe worth to merge after #1002 and expand tests

### CODE REVIEW

File: src/brevitas/core/stats/stats_op.py [hunk 1]
```diff
@@ -544,3 +546,216 @@ def forward(self, x):
...
+                mask = (num != 0.) & (den != 0.)
+                if self.stats_reduce_dim is None:
+                    candidate = masked_median(num / den, mask)
+                else:
+                    candidate = masked_median(num / den, mask, dim=self.stats_reduce_dim)
+                candidate = self.input_view_shape_impl(candidate)
+                candidate[torch.isnan(candidate)] = self.internal_candidate[torch.isnan(candidate)]
+                candidate = self.clamp_min_ste(candidate)
+                bit_width = self.msb_clamp_bit_width_impl()
+                int_threshold = self.int_scaling_impl(bit_width)
```
Comment (Giuseppe5) [inline src/brevitas/core/stats/stats_op.py:621]: Int specific implementation, maybe it should be generalized.

File: src/brevitas/quant/base.py [hunk 2]
```diff
@@ -502,3 +504,53 @@ class MSEWeightZeroPoint(MSEZeroPoint):
 
 class MSEActZeroPoint(MSEZeroPoint):
     zero_point_impl = ParameterFromRuntimeZeroPoint
+
+
+class HQOZeroPoint(ExtendedInjector):
+    """
+    We leverage a sub-injector to avoid a name clash between scale and zero-point.
```
Comment (Giuseppe5) [inline src/brevitas/quant/base.py:511]: Remove here and next

File: src/brevitas/core/stats/stats_op.py [hunk 3]
```diff
@@ -440,6 +442,21 @@ def _set_local_loss_mode(module, enabled):
             m.local_loss_mode = enabled
 
 
+def _set_quantize_zero_point(module, enabled):
```
Comment (Giuseppe5) [inline src/brevitas/core/stats/stats_op.py:445]: Remove

File: src/brevitas/core/stats/stats_op.py (hunk 1 above)
Comment (nickfraser) [reply src/brevitas/core/stats/stats_op.py:622]: This looks like it hasn't been resolved to me. Is that true?

**Review by nickfraser — COMMENTED** [review]
Is it intentional that this only works for Int?

**Pull request merged**